Listen for group member state changes when using `expand` in templates

Tho85 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

This PR amends the logic that determines what state changes should be listened for when using a template (sensor). Up until now, groups were not handled as expected.

Consider the minimal configuration example below.

_Expected behavior_: The sensor should always display the sum of enabled switches, i.e. if I enable the first switch, the sensor value changes from 0 to 1. If I then enable the second switch, the sensor changes from 1 to 2.

_Actual behavior_: The sensor value doesn't change at all when I toggle any of the switches.

This is due to the extraction of entities from templates that determines what state changes should be listened for. Up until now, the `expand` template function was not handled in that logic. This PR fixes this by expanding groups and returning the contained entities. To make this a non-breaking change, and because one might also listen for changes on the group (if there are any?), the group's entity ID will also be returned.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

input_boolean:
  switch_1:
    name: "Switch 1"
  switch_2:
    name: "Switch 2"

group:
  input_booleans:
    name: "All boolean inputs"
    entities:
      - input_boolean.switch_1
      - input_boolean.switch_2


sensor:
  - platform: template
    sensors:
      length_sensor:
        value_template: "{{ expand('group.input_booleans') | selectattr('state', 'eq', 'on') | list | length }}"
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Tho85 committed:
Re-evaluate template on group member state change

bdraco commented:
@Tho85 The change looks fine.  There is some coverage missing for the new code.

`pytest --cov=homeassistant/helpers/ --cov-report term-missing -- tests/helpers/test_template.py`
`homeassistant/helpers/template.py                     568     17    97%   57-64, 70, 100, 114, 206, 210-213, 255, 518`

bdraco reviewed (changes requested):
Please complete the test coverage on the new code.

Tho85 committed:
Use named groups for entity extraction regex

This will avoid unnecessary edits of match indices if the regex is
amended in the future

Tho85 committed:
Improve test coverage

Tho85 commented:
It looks like I happened to change some code that has been untested, by simply changing the regex match index in that line of code.

I added named groups to the regex to avoid this in the future, and I added some tests for the branches that were not yet tested.

bdraco reviewed (approved):
LGTM
